### DISCUSSION TITLE
Modeling the User entity

### DISCUSSION THREAD

**Discussion by JimmyTron** [SpaceyaTech/Team-Rio-Django#26] [Q&A] [answered]
### How do we efficiently modelling the User entity?
![Screenshot from 2022-11-06 13-38-58](https://user-images.githubusercontent.com/23496280/200166468-4ef0a58c-0371-4b3c-b907-91c4fe2a5659.png) 
[modeling the blog entities](https://docs.google.com/document/d/1KXtB83iaNLKG2HEtVhx5ZzoPz-DQDcycj2Wh7OX15-o/edit)

> from the properties here above from the first entity, our entity the user has several properties and youll agree that from using django some of the properties are primarily found on [the django user model ](https://docs.djangoproject.com/en/4.1/ref/contrib/auth/) 
> there are several ways in which we could implement the user on django;

- [ ] Extending the existing User model [read more](https://docs.djangoproject.com/en/4.1/topics/auth/customizing/#extending-the-existing-user-model)
- [x] Substituting a custom User model [read more](https://docs.djangoproject.com/en/4.1/topics/auth/customizing/#substituting-a-custom-user-model)
- [ ] Referencing the User model [read more](https://docs.djangoproject.com/en/4.1/topics/auth/customizing/#referencing-the-user-model)
- [ ] Specifying a custom user model [read more](https://docs.djangoproject.com/en/4.1/topics/auth/customizing/#specifying-a-custom-user-model)

**Which option from the ones above do you think would best work for this task?**

> all ideas are open for discussion.

**Comment by hellen-22** [comment, marked as answer]
Substituting a custom User model - it gives the chance to use the Django User Attributes(email, username, firstname, lastname, email), considering our model requires the Django user attributes. We will just need to add our extra fields(gender, account status, phone number e.t.c)

**Reply by JimmyTron** [reply]
assuming  we are entirely bound to use the combination of firstname and lastname as username.
![Screenshot from 2022-11-06 17-10-59](https://user-images.githubusercontent.com/23496280/200175694-b6877878-d0c3-4b9e-84de-c44ab52764bb.png)

![image](https://user-images.githubusercontent.com/23496280/200175682-5060ed48-9dde-4306-b856-6a6341cf8dca.png)

**Reply by JimmyTron** [reply]
![image](https://user-images.githubusercontent.com/23496280/200175796-edf7ad83-f54b-4cde-813a-da60ede25ade.png)

**Reply by JimmyTron** [reply]
@sangkips you could describe how we are to go about here too so that we could get started on the models

**Comment by sangkips** [comment]
What is the difference between substituting and extending the user model?

**Reply by wanjirumurira** [reply]
I think in this case, it means the same.

**Reply by JimmyTron** [reply]
Extending the existing User model 
![Screenshot_2022-11-07-14-10-02-791_com android chrome](https://user-images.githubusercontent.com/23496280/200296822-b7d62c58-3d18-467c-9bef-b0bce006ca34.png)

 Substituting a custom User model 
![Screenshot_2022-11-07-14-10-31-798_com android chrome](https://user-images.githubusercontent.com/23496280/200296885-e855be76-6ac3-4c0d-9c6b-9f08d53adb13.png)

**Reply by JimmyTron** [reply]
> I think in this case, it means the same.

Its not the same thing @wanjirumurira

**Reply by wanjirumurira** [reply]
Noted

**Reply by sangkips** [reply]
Can we manage users in a different application all together. Like we can have accounts app then we referenced it in the sheetings.py file with auth_user_model

**Reply by wanjirumurira** [reply]
Will we be using the username or an email to uniquely identify a user during authentication?

**Reply by sangkips** [reply]
Maybe we can do sign-in with either email or password of which if a user use username or email you will still login

**Reply by wanjirumurira** [reply]
From the documentation that Jimmy shared, it seems that one of the things that will guide us on which model to use is first deciding on how we'll want to uniquely identify a user during authentication.If we settle on username as the unique identifier, the default user model will work just fine but if we'd rather use an email address, extending or substituting(most preferred) the user model can work.

**Reply by Eugene-Kwaka** [reply]
From the backend meeting we had today, I think we will be extending the default user model since we will have the Account model that has a OneUserToManyAccounts relationship with the User. Since it is not yet confirmed if we will be identifying a user solely by their username or email, I think extending the default user model is enough. 

What do you guys think?

**Reply by hellen-22** [reply]
We will be extending the User model to the accounts Model. But then Django default user model does not have all the fields that we need our user to have. I guess we will need to customize the User Model first, then extend the customized model to the Accounts model.

**Reply by mamu234** [reply]
Hellen  please explain further how to  customize the User Model  and extend it  to the Accounts, I have gone through some documentation and am yet to get it

**Reply by sangkips** [reply]
I think here we are supposed to use oneToOnefield to create a relationship between the accounts and the user. Check out what have been working on and you can test. https://github.com/SpaceyaTech/Team-Rio-Django/discussions/26#discussioncomment-4107559

**Reply by hellen-22** [reply]
Customizing a user model is all about modelling a user model that meets your own user requirements that is different from the default model. The default django user model attributes are email, first_name, last_name, username, is_staff, is_active. Our user model needs all the default attributes plus other attributes like phone_number and verification_code. So instead of specifying a custom user(Which would require modelling the user model fully i.e. include all the fields even the available ones by default), we just substitute a custom user and add the extra fields in the following way:

```
from django.db import models
from django.contrib.auth.models import AbstractUser

class User(AbstractUser):
    email = models.EmailField(unique=True)
    phone_number = models.CharField(max_length=30, blank=True)
    verification_code = models.CharField(max_length=100)

    USERNAME_FIELD = 'email'
    REQUIRED_FIELDS = ['first_name', 'last_name', 'username']

    def __str__(self) -> str:
        return f'{self.first_name} {self.last_name}' 
```

<p>The AbstractUser gives the other fields and then you'll need to add the extra fields(phone_number, verification_code). I included email in the fields to make it unique. It is not unique by default. So I chose to make it unique.
The USERNAME_FIELD specifies what will be used for login. REQUIRED_FIELDS specifies the fields required when creating a superuser.

After customizing the user you need to update your seeting.py file and "point" to where the custom user is:

AUTH_USER_MODEL = 'accounts.User'

In this case the user model is in the accounts app.

The remaining part is extending the custom user model to the account: </p>

```
class Account(models.Model):
    user = models.ForeignKey(User, on_delete=models.CASCADE)
    name = models.CharField(max_length=50)
    created_at = models.DateTimeField(auto_now_add=True)
    updated_at = models.DateTimeField(auto_now=True)
    display_picture = models.ImageField()
    bio = models.TextField()

    def __str__(self) -> str:
        return self.name
```

The user is a foreign key on the account model, that's basically extending the user model to the accounts model.
With this you've created the custom user and now there's a relationship between the user and account.

That's the approach i have taken.

**Reply by mamu234** [reply]
Thanks, Hellen this is quite elaborate ,now I get it

**Reply by hellen-22** [reply]
@sangkips I think its a one to many relationship because a user can have several accounts. We use ForeignKeyField.

**Reply by sangkips** [reply]
@hellen-22 I think a user should have just one account. Let's take an example of Medium, yes you can have two or three accounts but with different user details, you can not authenticate a different user with the same email. The way you will bypass the system you as a human being is by using different social accounts but from the systems point of view it's two different accounts because of the verification tokens and handshakes done behind the scenes.
How I see the accounts model is, it holds the additional attributes of the user. Think of it as a profile, you register then you have an option to add additional information if you want. When you add those information, you are adding only to that instance of a user in the database you logged in.

**Reply by hellen-22** [reply]
We've been discussing the models in the Backend meetings with Hudson and it was decided that a user can have several accounts.
Here's the repo to the models: [https://github.com/SpaceyaTech/CMS-Backend-Repository]

If the Account model was holding additional attributes of the user we wouldn't need it because those attributes could just be added to the customized user model.

**Reply by Eugene-Kwaka** [reply]
I second what Hellen is suggesting. In this case The user will the extend to the accounts model with a OneToMany relationship since that is what he Backend Team agreed on. 

@sangkips Using Medium's way, you as a user will have to use different social accounts as different profiles to use the system. you will signup to multiple accounts just to create different profiles of yourself.

**Comment by victorazangu** [comment]
On Verification, are we verifying the user or account?

**Reply by JimmyTron** [reply]
Ideally authentication should be on the user. Since our ER Model has most of the unique identifiers on the user table. Another suggestion is to use a third party authentication for the application. More clarification is fo be sort on how the authentication will be done. @sangkips you still have to state your case on the same.

**Comment by sangkips** [comment]
Django built-in authentication system can serve all our needs very well, but sometimes we need to do some fine adjustments in order to fit our web application needs.
If the application has a social appeal which ours has, we might want to store a short bio, profile picture of the user, and other things like that.

For our case, we need to store the bio, and profile picture and reference it to our user. Check how WordPress does this. What we will do is, use a oneToOnefield relationship with our user model(account). We want the user to signup using an Email or username and a password only and once logged in one can edit the profile. Here the profile/account is not related to any authentication process.

We want to have an instance whereby when a user updated his profile picture or bio, the information is captured automatically. Read about signals in Django, here is a small hint on signals.
signals are utilities that help us to connect events with actions. We can develop a function that will run when a signal calls it. In other words, Signals are used to perform some action on modification/creation of a particular entry in the Database. For example, One would want to create a profile instance, as soon as a new user instance is created in the Database.

Here is how I will approach it.
## models.py file

```
from django.db import models
from django.contrib.auth.models import User

class Account(models.Model):
    user = models.OneToOneField(User, related_name='account', on_delete=models.CASCADE )
    bio = models.CharField(max_length=30, blank=True, null=True)
    #profile_picture = models.ImageField(default="", upload_to="You/media/uploads", blank=True)
    created_at = models.DateTimeField(auto_created=True)
    updated_at = models.DateTimeField(auto_now=True)

    def __str__(self):
        return self.name

STATUS = (
    (0,"Draft"),
    (1,"Publish")
)

class Blog(models.Model):
    title = models.CharField(max_length=200)
    blog_description = models.CharField(max_length=255)
    content = models.TextField()
    author = models.ForeignKey(Account, on_delete=models.CASCADE)
    status = models.IntegerField(choices=STATUS, default=0)
    created_at = models.DateTimeField(auto_created=True)
    updated_at = models.DateTimeField(auto_now= True)

    class Meta:
        ordering = ['-created_at']

    def __str__(self):
        return self.title
```
![Screenshot from 2022-11-10 16-34-32](https://user-images.githubusercontent.com/15798975/201105602-f300d9a9-0ca9-4543-9158-8a671ea55ff1.png)

**Reply by teslima0** [reply]
Substituting a custom User model -  this enable us to use the Django UserBase Attributes( such email, username, firstname, lastname, email) that are very important in the development. We only have to add other extra fields(gender, account status, phone number e.t.c) and customize it to our taste.

**Comment by JosephThuku** [comment]
inorder to implement the many to on relationship that is one user can has many accounts and one account can have many blogposts. we need to use foreign keys. for me i would use the inbuild user model and then for the account and blogpost this is how i would solve
![Screenshot (27)](https://user-images.githubusercontent.com/93326983/201659880-8ea8c1a2-04fc-4bef-81af-b91b88741ddb.png)
